### DISCUSSION TITLE
Integrating Fast-DDS into a project without installing external dependencies system wide.

### DISCUSSION THREAD

**Discussion by Renegade85** [eProsima/Fast-DDS#3849] [Q&A]
Hello,

we are using Fast-DDS library in our project. We would like to update our build system to CMake. A simplified structure of our project looks like this:

```
project
+-- CMakeLists.txt
+-- 3rdparty
    +--Fast-DDS
    +--foonathan_memory_vendor

```

The projects CMakeLists.txt file looks following way:

>   1 cmake_minimum_required (VERSION 3.16)
>   2
>   3 message(STATUS "default install: ${CMAKE_INSTALL_PREFIX}")
>   4
>   5 set (THIRDPARTY ON CACHE STRING "" FORCE)
>   6 set (BUILD_SHARED_LIBS OFF CACHE STRING "" FORCE)
>   7
>   8
>   9 add_subdirectory(3rdparty/foonathan_memory_vendor)
>  10 add_subdirectory(3rdparty/Fast-DDS)
>  11

I've tried to generate a build configuration, but this fails due to the fact, that Fast-DDS uses `find_package(foonathan_memory_vendor)`, however, that package is not installed in the system and I don't want to install it system wide. Unfortunately I can't figure out, how to make this running, as I'm a cmake newbie. Any help would be really appreciated.

Second question related to this is, why is the foonathan_memory_vendor used instead of the original foonathan_memory?
Why is the foonathan_memory "included" as an external dependency and is not part of the thirdparty directory as other libs?

Thanks a lot for your help

**Comment by JLBuenoLopez** [comment]
Hi @Renegade85 

[`foonathan_memory_vendor`](https://github.com/eProsima/foonathan_memory_vendor#foonathan-memory-vendor) is an auxiliary package to check if `foonathan_memory` is installed and if not, download and install it. If you look into the CMakeLists.txt in eProsima repository, what is required there is [`find_package(foonathan_memory)`](https://github.com/eProsima/Fast-DDS/blob/master/CMakeLists.txt#L239). Maybe you have been misled by Fast DDS installation guide. In those instructions, `foonathan_memory_vendor` is directly used instead of asking the user to install the package by their own. But you can remove the vendor dependency if you have already installed `foonathan_memory` in your system.

**Reply by Renegade85** [reply]
Hi @JLBuenoLopez-eProsima 

I see. The find_package( ) requires the module/package to be installed, is that correct? I'm using the add_subdirectory way to include things, so this won't work. Am I correct? Or can I use the add_subdirectory(foonathan_memory) (not the vendor one) and it will work without the need to install the package?

I'm sorry, I'm new to cmake and I still miss all the relationships between modules etc.

**Reply by JLBuenoLopez** [reply]
Fast DDS current configuration does not allow to include `foonathan_memory` with the `add_subdirectory` approach. This should be fixed including `foonathan_memory` in the thirdparty directory as you pointed in your opening question. You might want to look at #3818 which has a workaround to use `add_subdirectory` to include `foonathan_memory`.

**Reply by Renegade85** [reply]
Do I understand the issue #3818 that you plan to add the foonathan as a submodule under thirdparty or it was just a suggestion?

**Reply by JLBuenoLopez** [reply]
We will try to do it but I cannot promise a specific time frame.

**Reply by Renegade85** [reply]
Thanks a lot for your help
